### PR TITLE
ARROW-686: [C++] Account for time metadata changes, add Time32 and Time64 types

### PR DESCRIPTION
**Pull request by wesm**
This also has a little visitor refactoring

### CODE REVIEW

**Comment by wesm** [comment]
@xhochy I removed the copy ctors for `arrow::DataType` after running into a linker error on MSVC -- they weren't being used anywhere

File: cpp/src/arrow/ipc/json-internal.cc [hunk 1]
```diff
@@ -736,9 +762,10 @@ class JsonSchemaReader {
       // TODO
       *type = date64();
     } else if (type_name == "time") {
-      return GetTimeLike<TimeType>(json_type, type);
+      // TODO: 32-bit vs 64-bit time
```
Comment (xhochy) [inline cpp/src/arrow/ipc/json-internal.cc]: Still relevant? It looks like this is implemented.

**Review by xhochy — APPROVED** [review]
+1, code looks fine only the TODO seems obsolete.

File: cpp/src/arrow/ipc/json-internal.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: will remove, thanks. Planning to do some cleaning on this module while completing the JSON read/write support

**Comment by wesm** [comment]
I am going to fix glib compilation, remove the TODO, and then merge

**Comment by wesm** [comment]
This build passes except for parquet-cpp build failure. I filed PARQUET-923

**Pull request closed without merging**